File: lua/autorun/server/simfphys_trailers_reborn.lua
-- Lua Library inline imports
function __TS__ArrayFilter(arr, callbackfn)
    local result = {}
    do
        local i = 0
        while i < #arr do
            if callbackfn(_G, arr[i + 1], i, arr) then
                result[#result + 1] = arr[i + 1]
            end
            i = i + 1
        end
    end
    return result
end

function __TS__ArrayForEach(arr, callbackFn)
    do
        local i = 0
        while i < #arr do
            callbackFn(_G, arr[i + 1], i, arr)
            i = i + 1
        end
    end
end

function __TS__ArrayPush(arr, ...)
    local items = {...}
    for ____, item in ipairs(items) do
        arr[#arr + 1] = item
    end
    return #arr
end

local ____exports = {}
local Trailers
if not simfphys then
    error("TR: missing: simfphys (https://steamcommunity.com/workshop/filedetails/?id=771487490)")
end
AddCSLuaFile("autorun/client/TR_debugspheres.lua")
print("loads")
local function valid(safe)
    if safe and (safe > 255) then
        error("TR: stack overflow?")
    end
    Trailers.cars = __TS__ArrayFilter(
        Trailers.cars,
        function(____, ventity)
            return IsValid(ventity.ent)
        end
    )
    __TS__ArrayForEach(
        Trailers.cars,
        function(____, ventity)
            if ventity.connection and (not IsValid(ventity.connection.socket)) then
                ventity.connection = nil
            end
        end
    )
end
local function findVEntity(entity)
    do
        local i = 0
        while i < #Trailers.cars do
            if Trailers.cars[i + 1].ent == entity then
                return Trailers.cars[i + 1]
            end
            i = i + 1
        end
    end
    print("not found")
end
local function getNext(ventity)
    if ventity and ventity.connection then
        return findVEntity(ventity.connection.ent)
    end
end
local function getwhole(ventity)
    local buffer = {}
    local current = ventity
    __TS__ArrayPush(buffer, current)
    while getNext(current) ~= nil do
        current = getNext(current)
        __TS__ArrayPush(buffer, current)
    end
    return buffer
end
local function IsConnectable(vent1, vent2)
    if vent1.output and vent2.input then
        return vent1.ent:LocalToWorld(vent1.output):DistToSqr(
            vent2.ent:LocalToWorld(vent2.input)
        ) < 400
    end
end
local function GetConnectable(ventity)
    do
        local i = 0
        while i < #Trailers.cars do
            do
                if ventity == Trailers.cars[i + 1] then
                    goto __continue18
                end
                if IsConnectable(ventity, Trailers.cars[i + 1]) then
                    return Trailers.cars[i + 1]
                end
            end
            ::__continue18::
            i = i + 1
        end
    end
end
Trailers = {}
do
    Trailers.cars = {}
    Trailers.systems = {}
    function Trailers.Init(ventity)
        __TS__ArrayPush(Trailers.cars, ventity)
        net.Start("trailers_reborn_debug_spheres", true)
        net.WriteTable({ent = ventity.ent, input = ventity.input, output = ventity.output})
        net.Broadcast()
    end
    function Trailers.InitBySpawn(ent)
        if IsValid(ent) then
            local spawnname = ent:GetSpawn_List()
            if IsValid(ent) and (spawnname ~= "") then
                local spawnlistentity = list.Get("simfphys_vehicles")[spawnname]
                if spawnlistentity then
                    if spawnlistentity.Members and spawnlistentity.Members.Trailers then
                        spawnlistentity.Members.Trailers.ent = ent
                        Trailers.Init(spawnlistentity.Members.Trailers)
                    else
                        print(
                            "TR:\n\tvehicle's spawnlist missing 'Members' or 'Members.Trailers'\n" .. tostring(
                                debug.traceback()
                            )
                        )
                    end
                else
                    print(
                        "TR:\n\tvehicle is missing simfphys spawn list\n" .. tostring(
                            debug.traceback()
                        )
                    )
                end
            else
                print(
                    (((("TR:\n\t" .. tostring(ent)) .. "\n\t") .. tostring(spawnname)) .. "\n\t^ something isn't valid\n") .. tostring(
                        debug.traceback()
                    )
                )
            end
        else
            print(
                (("TR:\n\t" .. tostring(ent)) .. " is NOT valid\n") .. tostring(
                    debug.traceback()
                )
            )
        end
    end
    function Trailers.Connect(ventity)
        if ventity == nil then
            print("TR: ventity == null")
            print(
                debug.traceback()
            )
            return
        end
        local whole = getwhole(ventity)
        PrintTable(whole, 0, {})
        print(#whole)
        ventity = whole[#whole]
        local vtrailer = GetConnectable(ventity)
        if vtrailer then
            PrintTable(vtrailer, 0, {})
            local ballsocketent = constraint.AdvBallsocket(ventity.ent, vtrailer.ent, 0, 0, ventity.output, vtrailer.input, 0, 0, 0, 0, 0, 360, 360, 360, 0, 0, 0, 0, 0)
            ventity.connection = {ent = vtrailer.ent, socket = ballsocketent}
        else
            print("TR: no connectable trailers found :C")
        end
    end
    function Trailers.ConnectEnt(entity)
        valid()
        Trailers.Connect(
            findVEntity(entity)
        )
    end
    function Trailers.Disconnect(ventity)
        if ventity == nil then
            print("TR: ventity == null")
            print(
                debug.traceback()
            )
            return
        end
        if ventity.connection then
            local whole = getwhole(ventity)
            PrintTable(whole, 0, {})
            print(#whole)
            ventity = whole[(#whole - 2) + 1]
            SafeRemoveEntity(ventity.connection.socket)
            ventity.connection = nil
        else
            print("TR: no connections found")
            print(
                debug.traceback()
            )
        end
    end
    function Trailers.DisconnectEnt(entity)
        valid()
        Trailers.Disconnect(
            findVEntity(entity)
        )
    end
end
local files = ({
    file.Find("TR_extensions/*", "LUA")
})[1]
for ____, system in ipairs(files) do
    print(system)
    __TS__ArrayPush(
        Trailers.systems,
        include(
            "TR_extensions/" .. tostring(system)
        )
    )
end
timer.Remove("TR_system")
timer.Create(
    "TR_system",
    0.5,
    0,
    function()
        valid()
        for ____, ventity in ipairs(Trailers.cars) do
            for ____, system in ipairs(Trailers.systems) do
                if system.HandleTruck then
                    system.HandleTruck(ventity)
                end
            end
        end
    end
)
hook.Add(
    "OnEntityCreated",
    "TR_handle",
    function(ent)
        if ent:GetClass() == "gmod_sent_vehicle_fphysics_base" then
            timer.Simple(
                0.1,
                function()
                    Trailers.InitBySpawn(ent)
                end
            )
        end
    end
)
concommand.Add(
    "trailers_connect",
    function(ply)
        if IsValid(ply) then
            print(
                ply:GetSimfphys()
            )
            Trailers.ConnectEnt(
                ply:GetSimfphys()
            )
        else
            print("TR: IsValid(ply) == false")
            print(
                debug.traceback()
            )
        end
    end
)
concommand.Add(
    "trailers_disconnect",
    function(ply)
        if IsValid(ply) then
            print(
                ply:GetSimfphys()
            )
            Trailers.DisconnectEnt(
                ply:GetSimfphys()
            )
        else
            print("TR: IsValid(ply) == false")
            print(
                debug.traceback()
            )
        end
    end
)
util.AddNetworkString("trailers_reborn_debug_spheres")
_G.Trailers = Trailers
return ____exports

File: src/autorun/server/simfphys_trailers_reborn.ts
declare const simfphys: any
declare function error(this: void, message: string, errorLevel?: number): void
declare namespace debug { function traceback(this: void): string }

if (!simfphys) {
	error("TR: missing: simfphys (https://steamcommunity.com/workshop/filedetails/?id=771487490)")
}
AddCSLuaFile("autorun/client/TR_debugspheres.lua")
print("loads")
function valid(this: void, safe?: number) {
	if (safe && safe > 255) error("TR: stack overflow?")
	Trailers.cars = Trailers.cars.filter((ventity) => { return IsValid(ventity.ent) })
	Trailers.cars.forEach((ventity) => {
		if (ventity.connection && !IsValid(ventity.connection.socket)) {
			ventity.connection = null as any as VConnection
		}
	})
}
function findVEntity(this: void, entity: Entity) {
	for (let i = 0; i < Trailers.cars.length; i++) {
		if (Trailers.cars[i].ent == entity)
			return Trailers.cars[i]
	}
	print("not found")
}
function getNext(this: void, ventity: VEntity | undefined) {
	if (ventity && ventity.connection) return findVEntity(ventity.connection.ent)
}
function getwhole(this: void, ventity: VEntity) {
	let buffer: VEntity[] = []
	let current: VEntity | undefined = ventity
	buffer.push(current)
	while (getNext(current) != null) {
		//if (current != null)
		//buffer.push(current)
		current = getNext(current)
		buffer.push(current as VEntity)
	}
	return buffer
}
function IsConnectable(this: void, vent1: VEntity, vent2: VEntity) {
	//return true
	if (vent1.output && vent2.input)
		return vent1.ent.LocalToWorld(vent1.output).DistToSqr(vent2.ent.LocalToWorld(vent2.input)) < 400
}
function GetConnectable(this: void, ventity: VEntity) {
	for (let i = 0; i < Trailers.cars.length; i++) {
		if (ventity == Trailers.cars[i]) continue
		if (IsConnectable(ventity, Trailers.cars[i])) return Trailers.cars[i]
	}
}
namespace Trailers {
	export let cars: VEntity[] = []
	export const systems: System[] = []
	export function Init(this: void, ventity: VEntity) {
		Trailers.cars.push(ventity)
		net.Start("trailers_reborn_debug_spheres", true)
		net.WriteTable({ ent: ventity.ent, input: ventity.input, output: ventity.output })
		net.Broadcast()
	}
	export function InitBySpawn(this: void, ent: Entity | any) {
		if (IsValid(ent)) {
			const spawnname = ent.GetSpawn_List()
			if (IsValid(ent) && spawnname != "") {
				const spawnlistentity = list.Get("simfphys_vehicles")[spawnname]
				if (spawnlistentity) {
					if (spawnlistentity.Members && spawnlistentity.Members.Trailers) {
						spawnlistentity.Members.Trailers.ent = ent
						Trailers.Init(spawnlistentity.Members.Trailers)
					} else {
						print(`TR:
	vehicle's spawnlist missing 'Members' or 'Members.Trailers'
${debug.traceback()}`)
					}
				} else {
					print(`TR:
	vehicle is missing simfphys spawn list
${debug.traceback()}`)
				}
			} else {
				print(`TR:
	${ent}
	${spawnname}
	^ something isn't valid
${debug.traceback()}`)
			}
		} else {
			print(`TR:
	${ent} is NOT valid
${debug.traceback()}`)
		}
	}
	export function Connect(this: void, ventity: VEntity | undefined) {
		if (ventity == undefined) {
			print("TR: ventity == null")
			print(debug.traceback())
			return
		}
		const whole = getwhole(ventity)
		PrintTable(whole, 0, {})
		print(whole.length)
		ventity = whole[whole.length - 1]
		const vtrailer = GetConnectable(ventity)
		if (vtrailer) {
			PrintTable(vtrailer as VEntity, 0, {})
			const ballsocketent = constraint.AdvBallsocket(ventity.ent, vtrailer.ent, 0, 0, ventity.output, vtrailer.input, 0, 0, 0, 0, 0, 360, 360, 360, 0, 0, 0, 0, 0)
			ventity.connection = { ent: vtrailer.ent, socket: ballsocketent }
		} else {
			print("TR: no connectable trailers found :C")
		}
	}
	export function ConnectEnt(this: void, entity: Entity) {
		valid()
		Trailers.Connect(findVEntity(entity))
	}
	export function Disconnect(this: void, ventity: VEntity | undefined) {
		if (ventity == null) {
			print("TR: ventity == null")
			print(debug.traceback())
			return
		}
		if (ventity.connection) {
			const whole = getwhole(ventity)
			PrintTable(whole, 0, {})
			print(whole.length)
			ventity = whole[whole.length - 2]
			SafeRemoveEntity(ventity.connection.socket)
			// WHAT ?
			ventity.connection = null as any as VConnection
		} else {
			print("TR: no connections found")
			print(debug.traceback())
		}
	}
	export function DisconnectEnt(this: void, entity: Entity) {
		valid()
		Trailers.Disconnect(findVEntity(entity))
	}
}

const files = file.Find("TR_extensions/*", "LUA")[0]
for (const system of (files as string[])) {
	print(system)
	// still WHAT?
	Trailers.systems.push(include("TR_extensions/" + system) as any as System)
}
timer.Remove("TR_system")
timer.Create("TR_system", 0.5, 0, () => {
	valid()
	for (const ventity of Trailers.cars) {
		for (const system of Trailers.systems) {
			if (system.HandleTruck) {
				system.HandleTruck(ventity)
			}
		}
	}
})
hook.Add("OnEntityCreated", "TR_handle", (ent) => {
	if (ent.GetClass() == "gmod_sent_vehicle_fphysics_base") {
		timer.Simple(0.1, () => {
			Trailers.InitBySpawn(ent)
		})
	}
})
concommand.Add("trailers_connect", (ply: Player | any) => {
	if (IsValid(ply)) {
		print(ply.GetSimfphys())
		Trailers.ConnectEnt(ply.GetSimfphys())
	} else {
		print("TR: IsValid(ply) == false")
		print(debug.traceback())
	}
})
concommand.Add("trailers_disconnect", (ply: Player | any) => {
	if (IsValid(ply)) {
		print(ply.GetSimfphys())
		Trailers.DisconnectEnt(ply.GetSimfphys())
	} else {
		print("TR: IsValid(ply) == false")
		print(debug.traceback())
	}
})
util.AddNetworkString("trailers_reborn_debug_spheres")
_G["Trailers"]= Trailers
export { }
